### PR TITLE
Fix --split-max-size

### LINKED ISSUE #6634: Inconsistent size output of chunks with gguf-split
System:
Ubuntu server 22.04 LTS
5950X
64GB Ram 3200Mhz
2x Nvidia 3090

Steps to reproduce:

First I converted the model to FP16 GGUF with:
`./convert.py --outfile Karasu-Mixtral-8x22B-v0.1-fp16.gguf --outtype f16 lightblue_Karasu-Mixtral-8x22B-v0.1`

That worked just fine and I got:
image

Then to quantize it to Q5_K_M:
`./quantize Karasu-Mixtral-8x22B-v0.1-fp16.gguf Karasu-Mixtral-8x22B-v0.1-Q5_K_M.gguf Q5_K_M`

That worked fine too:
image

But when using gguf-split --split even though I'm using --split-max-tensors 128 the sizes of the chunks are inconsistent:

`./gguf-split --split --split-max-tensors 128 /nfs/models/Karasu-Mixtral-8x22B-v0.1-Q5_K_M.gguf /nfs/models/`

image
image

### LINKED ISSUE #6654: Trying to split model with --split-max-size, but gguf-split ignores it
Latest version, ubuntu 2204, conda python=3.10.
Trying to split model with gguf-split, but something is going wrong 
```
(base) richard@richard-ProLiant-DL580-Gen9:~/Desktop/ramdisk/banana/llama.cpp$ ./gguf-split --split --split-max-size 4000M --dry-run /media/richard/5fbd0bfa-8253-4803-85eb-80a13218a927/grok-1-fp16-gguf/grok-1-Q5_K.gguf Q5_K/grok-1 
n_split: 1
split 00001: n_tensors = 2115, total_size = 214437M
gguf_split: 1 gguf split written with a total of 2115 tensors.
(base) richard@richard-ProLiant-DL580-Gen9:~/Desktop/ramdisk/banana/llama.cpp$ ./gguf-split --split --split-max-size 4G --dry-run /media/richard/5fbd0bfa-8253-4803-85eb-80a13218a927/grok-1-fp16-gguf/grok-1-Q5_K.gguf Q5_K/grok-1 
n_split: 17
split 00001: n_tensors = 128, total_size = 14609M
split 00002: n_tensors = 128, total_size = 13184M
split 00003: n_tensors = 128, total_size = 12648M
split 00004: n_tensors = 128, total_size = 12597M
split 00005: n_tensors = 128, total_size = 12648M
split 00006: n_tensors = 128, total_size = 12750M
split 00007: n_tensors = 128, total_size = 12836M
split 00008: n_tensors = 128, total_size = 13088M
split 00009: n_tensors = 128, total_size = 13197M
split 00010: n_tensors = 128, total_size = 12597M
split 00011: n_tensors = 128, total_size = 12597M
split 00012: n_tensors = 128, total_size = 12699M
split 00013: n_tensors = 128, total_size = 12699M
split 00014: n_tensors = 128, total_size = 12597M
split 00015: n_tensors = 128, total_size = 13137M
split 00016: n_tensors = 128, total_size = 13675M
split 00017: n_tensors = 67, total_size = 6868M
gguf_split: 17 gguf split written with a total of 2115 tensors.
```

### PR DESCRIPTION
**Pull request by CISC**
Byte size calculation was done on int and overflowed.

Fixes #6634
Fixes #6654

### CODE REVIEW

**Review by phymbert — CHANGES_REQUESTED** [review]
Thanks, but we discussed in #6343 that taking into account the wide usage of this feature to cover it with tests to avoid reproduce it in the future.

I do not want to block this critical fix, but would you please include [tests.sh](https://github.com/phymbert/llama.cpp/blob/hp/split/max-size/examples/gguf-split/tests.sh) and add it to the CI.

If you have not the time, I can do it later today.

**Comment by CISC** [comment]
> I do not want to block this critical fix, but would you please include [tests.sh](https://github.com/phymbert/llama.cpp/blob/hp/split/max-size/examples/gguf-split/tests.sh) and add it to the CI.

Sure, I'll look into it now.

**Review by phymbert — APPROVED** [review]
Thanks. FYI @ngxson

**Comment by CISC** [comment]
Ok, done. :)

I made some minor changes to tests.sh:
* Added a second optional parameter to set output directory
* Disabled the --no-tensor-in-metadata related tests (not working)
* Downloads a larger model and tests --split-max-size 2G instead of 40M

I've added autodiscovery of examples/*/tests.sh scripts (at that level only, so as not to include f.ex. examples/server/tests/tests.sh) to ci run. It will pass the bin and mnt directories as arguments. If you want to change how this works, I'm open to suggestions.

**Comment by phymbert** [comment]
Cool, the `--no-tensor-in-metadata` was for:

- #6463

But yet to be supported.

**Comment by ngxson** [comment]
The change to `gguf-split.cpp` looks good. Thanks for taking time on this.

However, for the test, I'm wondering:
- I couldn't find any workflow file refers to either `ci/run.sh` or `ci-run.sh`. I can either find any log for that on the workflow of this PR (or maybe I'm missing something). Can you confirm if the test have really been run on CI? @CISC 
- Having a complete test is always a good thing, but what hold me back atm is that we may accidentally make the workflow slower, takes more time to finish. Depending on the time it takes to finish the test for `gguf-split`, we may need to decide if we can keep using the current model (gemma Q8_0) or using the smaller one.

**Comment by CISC** [comment]
>     * I couldn't find any workflow file refers to either `ci/run.sh` or `ci-run.sh`. I can either find any log for that on the workflow of this PR (or maybe I'm missing something). Can you confirm if the test have really been run on CI? @CISC

To quote ci/README.md:

> In addition to [Github Actions](https://github.com/ggerganov/llama.cpp/actions) llama.cpp uses a custom CI framework:
> 
> https://github.com/ggml-org/ci
> 
> It monitors the master branch for new commits and runs the [ci/run.sh](https://github.com/ggerganov/llama.cpp/blob/master/ci/run.sh) script on dedicated cloud instances. This allows us to execute heavier workloads compared to just using Github Actions. Also with time, the cloud instances will be scaled to cover various hardware architectures, including GPU and Apple Silicon instances.
> 
> Collaborators can optionally trigger the CI run by adding the ggml-ci keyword to their commit message. Only the branches of this repo are monitored for this keyword.
> 
> It is a good practice, before publishing changes to execute the full CI locally on your machine

Which is what I did. :)

>     * Having a complete test is always a good thing, but what hold me back atm is that we may accidentally make the workflow slower, takes more time to finish. Depending on the time it takes to finish the test for `gguf-split`, we may need to decide if we can keep using the current model (gemma Q8_0) or using the smaller one.

The longest part of the test is downloading the model the first time, the next longest part is running the model tests, but I tested it with CPU build, and it's not really that slow.

**Comment by ngxson** [comment]
> To quote ci/README.md

IMO the goal of this test script is to test if your changes works before merging. However, what `ci/run.sh` does is test **after** it's merged into master branch, so it sounds a bit strange to me.

@phymbert It's up to you to decide if this behavior is expected or not.

> The longest part of the test is downloading the model the first time, the next longest part is running the model tests, but I tested it with CPU build, and it's not really that slow.

CPU in CI machines can be different from personal machines. For that reason, the workflow of server example uses the small tinyllama model (less than 100MB in size).

What I wanted to do earlier is to add the test and see if workflows in [gitlab actions](https://github.com/ggerganov/llama.cpp/actions) takes significantly longer to finish or not.

**Comment by CISC** [comment]
@ngxson Yeah, it is indeed a little strange, but according to the readme it's for exactly that purpose (not slowing down GHA)...

Just had a cursory glance at the CI repo and it looks like it preloads its image with ggml-org/models:phi-2/ggml-model-f16.gguf which is 5GB, perhaps that would be better to test against?

**Comment by phymbert** [comment]
the `ci\run.sh` is launched by `ggml/ci` manually by @ggerganov on Azure nodes, these tests aim to be slow, so adding a split tests is fine there IMHO.

It is OK if it is not run on all commits, we might ask any futur contributor to run it locally before merging.

But better to request an approval.

**Comment by ngxson** [comment]
I think the main problem is that we don't have the idea of running CI tests for `examples/*` before this PR (except for `server`). One idea that I have in my mind was that we can add a dedicated workflow for testing examples. But for now I'm OK with simply running `tests.sh` locally - I actually prefer to this way to keep things simple.

That's just my personal thought, maybe @ggerganov will have other ideas.

File: ci/run.sh [hunk 1]
```diff
@@ -153,6 +153,52 @@ function gg_sum_ctest_release {
...
+}
+
+# test_scripts_release
+
+function gg_run_test_scripts_release {
+    cd ${SRC}
+
+    set -e
+
+    (find ./examples -mindepth 2 -maxdepth 2 -name 'tests.sh' -execdir time bash "{}" "$SRC/build-ci-release/bin" "$MNT/models" \;) 2>&1 | tee -a $OUT/${ci}-scripts.log
```
Comment (ggerganov) [inline ci/run.sh:186]: Let's list explicitly the scripts that we run here - it's more clear what is being executed and easier to turn on/off certain tests

**Review by ggerganov — APPROVED** [review]
For `ggml-ci` we currently aim the runs to not take longer than 30 mins:

https://github.com/ggml-org/ci/blob/b2856375b21aa7d96bcc1d3a25d70f4446a0610d/env.sh#L113

Technically, this particular test does not need to run on more than one node since it does not reach any code that is hardware-dependent. But we can improve this in the future - let's merge like this for now

> I think the main problem is that we don't have the idea of running CI tests for examples/* before this PR (except for server).

We can add instructions when opening a PR that reminds contributors to run the `ggml-ci` CI locally

File: ci/run.sh (hunk 1 above)
Comment (CISC) [reply ci/run.sh]: Done

**Pull request merged**